Improved Kernel#raise

christopheraue opened the pull request:
1) ~~`raise` with no arguments now raises a NotImplementedError with a proper message.~~ (removed, see below)
2) ~~Re-raised errors now keep their backtrace like in CRuby.~~ (already added by #3840 in the meantime)
3) `raise exc, msg, backtrace` with a backtrace as third argument is now supported.

matz commented:
Unfortunately AppVeyor build failed.

christopheraue commented:
Yep, all builds failed. Wasn't considering running tests in a non-debug environment for a start...

christopheraue commented:
Builds with Linux, OSX and Visual Studio 14 are now fixed. But there is still a kind of strange failing test for Visual Studio 12. Despite the fact that it reports equal things as not equal the changes here do not have anything to do with mruby-eval. Any ideas?

ksss commented:
Why do you chose NotImplementedError instead of RuntimeError when `raise` with no arguments?
It seems different from ISO definition and CRuby behavior.

christopheraue commented on test/assert.rb:
```diff
@@ -166,6 +166,35 @@ def assert_raise(*exc)
   end
 end
 
+def assert_raise_msg(exc_cls, exc_msg)
```
Yes, I'll also include the `msg` parameter then.

christopheraue commented:
Because I've looked at it from the re-raise perspective only:

```
begin
  raise StandardError, 'error'
rescue
  raise
end
```

This is not supported by mruby. I've actually had this in my mruby code and wondered where the hell the RuntimeError without message came from. With no hint, that this is an mruby limitation. With that in mind I simply forgot that just a `raise` outside a rescue block should raise a RuntimeError without a message.

I'll differentiate between a `raise` inside and outside a rescue block. I just have to figure out how...

christopheraue commented:
Distinguishing between the inside and outside of rescue blocks involves too much work for now. I removed the commit that changed the behavior of `raise` without arguments.

`assert_raise_msg` has been renamed to `assert_raise_with_message`.

Visual Studio 12 is still complaining.

Asmod4n commented:
eval is currently broken (i believe), see #3817

christopheraue commented:
With the recent changes on master and #3817 fixed I rebased this PR on the latest master. All tests pass now.

Also, #3840 already fixed keeping backtraces in the meantime so I removed my changes in this regard.
